File: Node-js/2024/2024-006/2024-006.js
// guardPatrolLoopDetector.js

const fs = require('fs');
const path = require('path');

// Direction mappings
const DIRECTION_MAP = {
    '^': 0,
    '>': 1,
    'v': 2,
    '<': 3
};

const DIRECTION_OFFSETS = [
    [-1, 0], // Up
    [0, 1],  // Right
    [1, 0],  // Down
    [0, -1]  // Left
];

/**
 * Helper class to represent a position in the grid.
 */
class Position {
    constructor(row, col) {
        this.row = row;
        this.col = col;
    }

    toString() {
        return `${this.row},${this.col}`;
    }
}

/**
 * Helper class to represent a state (position and direction) of the guard.
 */
class State {
    constructor(row, col, direction) {
        this.row = row;
        this.col = col;
        this.direction = direction;
    }

    toString() {
        return `${this.row},${this.col},${this.direction}`;
    }
}

/**
 * Simple Pair class to hold two related objects.
 */
class Pair {
    constructor(first, second) {
        this.first = first;
        this.second = second;
    }
}

/**
 * Parses the grid from the given file.
 *
 * @param {string} filePath Path to the input file.
 * @returns {Array<Array<string>>} 2D character array representing the grid.
 */
function parseGrid(filePath) {
    const gridList = [];
    const fileContent = fs.readFileSync(filePath, 'utf-8');
    const lines = fileContent.split(/\r?\n/);

    for (let line of lines) {
        line = line.trim();
        if (line.length > 0) {
            gridList.push(line.split(''));
        }
    }

    if (gridList.length === 0) {
        throw new Error("The grid is empty.");
    }

    const cols = gridList[0].length;
    for (let row of gridList) {
        if (row.length !== cols) {
            throw new Error("Inconsistent row lengths in the grid.");
        }
    }

    return gridList;
}

/**
 * Finds the guard's starting position and direction.
 *
 * @param {Array<Array<string>>} grid 2D character array representing the grid.
 * @returns {Pair} A Pair containing the starting Position and direction.
 */
function findGuard(grid) {
    for (let r = 0; r < grid.length; r++) {
        for (let c = 0; c < grid[0].length; c++) {
            const cell = grid[r][c];
            if (DIRECTION_MAP.hasOwnProperty(cell)) {
                const guardPos = new Position(r, c);
                const guardDir = DIRECTION_MAP[cell];
                grid[r][c] = '.'; // Clear the starting position
                return new Pair(guardPos, guardDir);
            }
        }
    }
    throw new Error("Guard not found in the grid.");
}

/**
 * Retrieves all possible obstruction positions excluding the guard's starting position and already obstructed cells.
 *
 * @param {Array<Array<string>>} grid 2D character array representing the grid.
 * @param {Position} guardPos The starting position of the guard.
 * @returns {Array<Position>} List of possible obstruction Positions.
 */
function getPossibleObstructions(grid, guardPos) {
    const possible = [];
    for (let r = 0; r < grid.length; r++) {
        for (let c = 0; c < grid[0].length; c++) {
            if ((r !== guardPos.row || c !== guardPos.col) && grid[r][c] === '.') {
                possible.push(new Position(r, c));
            }
        }
    }
    return possible;
}

/**
 * Simulates the guard's movement on the grid.
 *
 * @param {Array<Array<string>>} grid 2D character array representing the grid.
 * @param {Position} startPos Starting position of the guard.
 * @param {number} startDir Starting direction of the guard.
 * @returns {boolean} True if a loop is detected, False if the guard exits the grid.
 */
function simulateMovement(grid, startPos, startDir) {
    const visitedStates = new Set();
    let r = startPos.row;
    let c = startPos.col;
    let direction = startDir;

    while (true) {
        const currentState = new State(r, c, direction).toString();
        if (visitedStates.has(currentState)) {
            return true; // Loop detected
        }
        visitedStates.add(currentState);

        const [dr, dc] = DIRECTION_OFFSETS[direction];
        const newR = r + dr;
        const newC = c + dc;

        // Check boundaries
        if (newR < 0 || newR >= grid.length || newC < 0 || newC >= grid[0].length) {
            return false; // Guard exits the grid
        }

        if (grid[newR][newC] === '#') {
            // Turn right if obstacle ahead
            direction = (direction + 1) % 4;
        } else {
            // Move forward
            r = newR;
            c = newC;
        }
    }
}

/**
 * Counts the number of distinct positions visited by the guard without any obstructions.
 *
 * @param {Array<Array<string>>} grid 2D character array representing the grid.
 * @param {Position} guardPos Starting position of the guard.
 * @param {number} guardDir Starting direction of the guard.
 * @returns {number} Number of distinct positions visited.
 */
function countDistinctPositionsVisited(grid, guardPos, guardDir) {
    const visitedPositions = new Set();
    visitedPositions.add(guardPos.toString());

    let r = guardPos.row;
    let c = guardPos.col;
    let direction = guardDir;

    while (true) {
        const [dr, dc] = DIRECTION_OFFSETS[direction];
        const newR = r + dr;
        const newC = c + dc;

        // Check boundaries
        if (newR < 0 || newR >= grid.length || newC < 0 || newC >= grid[0].length) {
            break; // Guard exits the mapped area
        }

        if (grid[newR][newC] === '#') {
            // Turn right if obstacle ahead
            direction = (direction + 1) % 4;
        } else {
            // Move forward
            r = newR;
            c = newC;
            visitedPositions.add(new Position(r, c).toString());
        }
    }

    return visitedPositions.size;
}

/**
 * Counts the number of obstruction positions that cause the guard to loop indefinitely.
 * Also measures and prints execution times.
 *
 * @param {Array<Array<string>>} grid 2D character array representing the grid.
 * @param {Position} guardPos Starting position of the guard.
 * @param {number} guardDir Starting direction of the guard.
 */
function countObstructionPositions(grid, guardPos, guardDir) {
    const totalStartTime = process.hrtime();

    // Clone the grid to prevent modifications
    const gridClone = grid.map(row => row.slice());

    // Time to find obstruction positions
    const obstructionStartTime = process.hrtime();
    const possibleObstructions = getPossibleObstructions(gridClone, guardPos);
    const obstructionEndTime = process.hrtime(obstructionStartTime);
    const obstructionTime = obstructionEndTime[0] + obstructionEndTime[1] / 1e9;

    // Print the first header and line: [time_obstruction_positions] [total_obstruction_positions]
    console.log("time, denominator");
    console.log(`${obstructionTime.toFixed(9)} ${possibleObstructions.length}`);

    // Print header for batches
    console.log("batch, batch time, cumulative time");

    let loopCount = 0;
    const total = possibleObstructions.length;
    const batchSize = 1000;
    let batchStartTime = process.hrtime();
    let cumulativeTime = obstructionTime;

    for (let idx = 0; idx < possibleObstructions.length; idx++) {
        const obstruction = possibleObstructions[idx];
        gridClone[obstruction.row][obstruction.col] = '#'; // Place obstruction

        if (simulateMovement(gridClone, guardPos, guardDir)) {
            loopCount++;
        }

        gridClone[obstruction.row][obstruction.col] = '.'; // Remove obstruction

        // Check if batch size is reached or it's the last position
        if ((idx + 1) % batchSize === 0 || (idx + 1) === total) {
            const batchEndTime = process.hrtime(batchStartTime);
            const batchTime = batchEndTime[0] + batchEndTime[1] / 1e9;
            cumulativeTime += batchTime;
            console.log(`${idx + 1} ${batchTime.toFixed(9)} ${cumulativeTime.toFixed(9)}`);
            batchStartTime = process.hrtime(); // Reset batch start time
        }
    }

    const totalEndTime = process.hrtime(totalStartTime);
    const totalTime = totalEndTime[0] + totalEndTime[1] / 1e9;

    // Print final answer header and line: [answer] [answer_time]
    console.log("answer, answer time");
    console.log(`${loopCount} ${totalTime.toFixed(9)}`);
}

/**
 * Main function to execute the Guard Patrol Loop Detector.
 */
function main() {
    // Specify the path to your input file here
    const filePath = path.join('\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-006\\input.txt');

    try {
        // Parse the grid
        const grid = parseGrid(filePath);

        // Find the guard's starting position and direction
        const guardInfo = findGuard(grid);
        const guardPos = guardInfo.first;
        const guardDir = guardInfo.second;

        // Part 1: Count distinct positions visited without obstructions
        const distinctPositions = countDistinctPositionsVisited(grid, guardPos, guardDir);
        console.log(`Number of distinct positions visited: ${distinctPositions}`);

        // Part 2: Detect loops with obstructions and measure execution times
        countObstructionPositions(grid, guardPos, guardDir);
    } catch (error) {
        console.error(`Error: ${error.message}`);
    }
}

// Execute the main function
main();

File: Scala-002/2024/2024-006/2024-006.scala
package com.peter_burbery.solve_advent_of_code_with_java_002.year_2024.day_006

import scala.io.Source
import scala.util.{Using, Try}
import scala.collection.mutable
import scala.collection.immutable.{Map => ImmutableMap}
import scala.util.control.Breaks.{break, breakable}

/**
 * GuardPatrolLoopDetector object to detect loops in a guard's patrol on a grid.
 */
object GuardPatrolLoopDetector {

  // Direction mappings
  private val DIRECTION_MAP: ImmutableMap[Char, Int] = ImmutableMap(
    '^' -> 0,
    '>' -> 1,
    'v' -> 2,
    '<' -> 3
  )

  private val DIRECTION_OFFSETS: Array[(Int, Int)] = Array(
    (-1, 0), // Up
    (0, 1),  // Right
    (1, 0),  // Down
    (0, -1)  // Left
  )

  def main(args: Array[String]): Unit = {
    val filePath = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-006\\input.txt" // Specify the path to your input file here

    try {
      // Part 1: Count distinct positions visited without obstructions
      val distinctPositions = countDistinctPositionsVisited(filePath)
      println(s"Number of distinct positions visited: $distinctPositions")

      // Part 2: Detect loops with obstructions and measure execution times
      countObstructionPositions(filePath)
    } catch {
      case e: Exception =>
        System.err.println(s"Error: ${e.getMessage}")
    }
  }

  /**
   * Part 1: Counts the number of distinct positions visited by the guard without any obstructions.
   *
   * @param filePath Path to the input file.
   * @return Number of distinct positions visited.
   */
  private def countDistinctPositionsVisited(filePath: String): Int = {
    // Parse the grid
    val grid = parseGrid(filePath)

    // Find the guard's starting position and direction
    val (guardPos, guardDir) = findGuard(grid)

    // Initialize visited positions set
    val visitedPositions: mutable.Set[Position] = mutable.Set(guardPos)

    var currentPos = guardPos
    var currentDir = guardDir

    // Simulate the guard's movement with breakable
    breakable {
      while (true) {
        val (dr, dc) = DIRECTION_OFFSETS(currentDir)
        val newR = currentPos.row + dr
        val newC = currentPos.col + dc

        // Check boundaries
        if (newR < 0 || newR >= grid.length || newC < 0 || newC >= grid(0).length) {
          break
        }

        if (grid(newR)(newC) == '#') {
          // Turn right if obstacle ahead
          currentDir = (currentDir + 1) % 4
        } else {
          // Move forward
          currentPos = Position(newR, newC)
          visitedPositions += currentPos
        }
      }
    }

    // Number of distinct positions visited
    visitedPositions.size
  }

  /**
   * Part 2: Counts the number of obstruction positions that cause the guard to loop indefinitely.
   * Also measures and prints execution times.
   *
   * @param filePath Path to the input file.
   */
  private def countObstructionPositions(filePath: String): Unit = {
    // Start total timing
    val totalStartTime = System.nanoTime()

    // Parse the grid
    val originalGrid = parseGrid(filePath)

    // Find the guard's starting position and direction
    val (guardPos, guardDir) = findGuard(originalGrid)

    // Time to find obstruction positions
    val obstructionStartTime = System.nanoTime()
    val possibleObstructions = getPossibleObstructions(originalGrid, guardPos)
    val obstructionEndTime = System.nanoTime()
    val obstructionTime = (obstructionEndTime - obstructionStartTime) / 1_000_000_000.0

    // Print the first header and line: [time_obstruction_positions] [total_obstruction_positions]
    println("time, denominator")
    println(f"$obstructionTime%.9f ${possibleObstructions.size}")

    // Print header for batches
    println("batch, batch time, cumulative time")

    // Initialize loop counter
    var loopCount = 0
    val total = possibleObstructions.size

    // Initialize timing for batches
    val batchSize = 1000
    var batchStartTime = System.nanoTime()
    var cumulativeTime = obstructionTime // cumulative_time includes obstruction_time

    for ((obstruction, idx) <- possibleObstructions.zipWithIndex) {
      // Create a mutable copy of the grid
      val grid = originalGrid.map(_.clone())
      grid(obstruction.row)(obstruction.col) = '#' // Place obstruction

      if (simulateMovement(grid, guardPos, guardDir)) {
        loopCount += 1 // Found a position that causes a loop
      }

      // Check if batch size is reached or it's the last position
      if ((idx + 1) % batchSize == 0 || (idx + 1) == total) {
        val batchEndTime = System.nanoTime()
        val batchTime = (batchEndTime - batchStartTime) / 1_000_000_000.0
        cumulativeTime += batchTime
        println(f"${idx + 1}%d $batchTime%.9f $cumulativeTime%.9f")
        batchStartTime = System.nanoTime() // Reset batch start time
      }
    }

    // End total timing
    val totalEndTime = System.nanoTime()
    val totalTime = (totalEndTime - totalStartTime) / 1_000_000_000.0 // Total time from start to end

    // Print final answer header and line: [answer] [answer_time]
    println("answer, answer time")
    println(f"$loopCount%d $totalTime%.9f")
  }

  /**
   * Parses the grid from the given file.
   *
   * @param filePath Path to the input file.
   * @return 2D character array representing the grid.
   */
  private def parseGrid(filePath: String): Array[Array[Char]] = {
    val gridList = mutable.ArrayBuffer[Array[Char]]()

    Using.resource(Source.fromFile(filePath)) { source =>
      for (line <- source.getLines()) {
        gridList += line.trim.toCharArray
      }
    }

    if (gridList.isEmpty) {
      throw new IllegalArgumentException("The grid is empty.")
    }

    // Ensure all rows have the same length
    val cols = gridList.head.length
    if (!gridList.forall(_.length == cols)) {
      throw new IllegalArgumentException("Inconsistent row lengths in the grid.")
    }

    gridList.toArray
  }

  /**
   * Finds the guard's starting position and direction.
   *
   * @param grid 2D character array representing the grid.
   * @return A tuple containing the starting Position and direction.
   */
  private def findGuard(grid: Array[Array[Char]]): (Position, Int) = {
    grid.zipWithIndex.flatMap { case (row, r) =>
      row.zipWithIndex.collect {
        case (cell, c) if DIRECTION_MAP.contains(cell) => (Position(r, c), DIRECTION_MAP(cell))
      }
    }.headOption match {
      case Some((guardPos, guardDir)) =>
        grid(guardPos.row)(guardPos.col) = '.' // Clear the starting position
        (guardPos, guardDir)
      case None =>
        throw new IllegalArgumentException("Guard not found in the grid.")
    }
  }

  /**
   * Retrieves all possible obstruction positions excluding the guard's starting position and already obstructed cells.
   *
   * @param grid     2D character array representing the grid.
   * @param guardPos The starting position of the guard.
   * @return List of possible obstruction Positions.
   */
  private def getPossibleObstructions(grid: Array[Array[Char]], guardPos: Position): List[Position] = {
    val possible = mutable.ListBuffer[Position]()
    for {
      r <- grid.indices
      c <- grid(r).indices
      if (r != guardPos.row || c != guardPos.col) && grid(r)(c) == '.'
    } {
      possible += Position(r, c)
    }
    possible.toList
  }

  /**
   * Simulates the guard's movement on the grid.
   *
   * @param grid     2D character array representing the grid.
   * @param startPos Starting position of the guard.
   * @param startDir Starting direction of the guard.
   * @return True if a loop is detected, False if the guard exits the grid.
   */
  private def simulateMovement(grid: Array[Array[Char]], startPos: Position, startDir: Int): Boolean = {
    val visitedStates: mutable.Set[State] = mutable.Set()
    var r = startPos.row
    var c = startPos.col
    var direction = startDir

    while (true) {
      val currentState = State(r, c, direction)
      if (visitedStates.contains(currentState)) {
        return true // Loop detected
      }
      visitedStates += currentState

      val (dr, dc) = DIRECTION_OFFSETS(direction)
      val newR = r + dr
      val newC = c + dc

      // Check boundaries
      if (newR < 0 || newR >= grid.length || newC < 0 || newC >= grid(0).length) {
        return false // Guard exits the grid
      }

      if (grid(newR)(newC) == '#') {
        // Turn right if obstacle ahead
        direction = (direction + 1) % 4
      } else {
        // Move forward
        r = newR
        c = newC
      }
    }

    false
  }

  /**
   * Case class to represent a position in the grid.
   *
   * @param row Row index.
   * @param col Column index.
   */
  private case class Position(row: Int, col: Int)

  /**
   * Case class to represent a state (position and direction) of the guard.
   *
   * @param row       Row index.
   * @param col       Column index.
   * @param direction Direction index.
   */
  private case class State(row: Int, col: Int, direction: Int)

  // Scala does not require a Pair class as tuples can be used instead.
}
